### PR TITLE
Document histograms and histogram_quantile.

### PR DESCRIPTION
**Pull request by beorn7**
@juliusv @brian-brazil

### CODE REVIEW

File: content/docs/practices/instrumentation.md [hunk 1]
```diff
@@ -206,11 +206,55 @@ Gauges can be set, go up, and go down. They are useful for snapshots of state,
...
-aggregatable](http://latencytipoftheday.blogspot.de/2014/06/latencytipoftheday-you-cant-average.html).
+### Summary vs. histogram
+
+Summaries and histograms are more complex metric types. They both sample
+observations. They track the number of observations *and* the sum of the
+observed values, allowing you to calculate the average observed value (useful
+for latency, for example). Note that the number of observations (showing up in
+Prometheus as a time series with a `_count` suffix) is inherently a counter (as
+described above, it only goes up), while the sum of observations (showing up as
+a time series with a `_sum` suffix) is inherently a gauge (if a negative value
```
Comment (brian-brazil) [inline content/docs/practices/instrumentation.md]: This is a bit messy, the cases where it won't be a counter are very small - and in that case the _sum is of questionable use.

File: content/docs/practices/instrumentation.md [hunk 2]
```diff
@@ -206,11 +206,55 @@ Gauges can be set, go up, and go down. They are useful for snapshots of state,
...
+server side using the [`histogram_quantile()`
+function](/docs/querying/functions/#histogram_quantile()).
+
+Both approaches have specific advantages and disadvantages:
+
+|   | Histogram | Summary
+|---|-----------|---------
+| Configuration | Need to configure buckets suitable for the expected range of observed values. | Need to configure φ-quantiles and sliding window, other φ-quantiles and sliding windows cannot be calculated later.
+| Client performance | Observations are very cheap as they only need to increment counters. | Observations are expensive due to the streaming quantile calculation.
+| Server performance | Calculating quantiles is expensive, consider [recording rules](/docs/querying/rules/#recording-rules) as a remedy. | Very low resource needs.
```
Comment (brian-brazil) [inline content/docs/practices/instrumentation.md]: Do we know how expensive calculating quantiles is?

File: content/docs/practices/instrumentation.md [hunk 3]
```diff
@@ -206,11 +206,55 @@ Gauges can be set, go up, and go down. They are useful for snapshots of state,
...
+| Number of time series | Low for Apdex score (see below), very high for accurate quantiles. Each bucket creates a time series. | Low, one time series per configured quantile.
+| Accuracy | Depends on number and layout of buckets. Higher accuracy requires more time series. | Configurable. Higher accuracy requires more client resources but is relatively cheap.
+| Specification of φ-quantile and sliding time window | Ad-hoc in Prometheus expressions. | Preconfigured by the client.
+| Aggregation | Ad-hoc aggregation with [Prometheus expressions](/docs/querying/functions/#histogram_quantile()). | In general [not aggregatable](http://latencytipoftheday.blogspot.de/2014/06/latencytipoftheday-you-cant-average.html).
+
```
Comment (brian-brazil) [inline content/docs/practices/instrumentation.md]: I think we should make mention that in practice having exact quantile values isn't necessary for most use cases, having an indication of which direction it's moving in is the more useful aspect and that can be at a much lower precision.

File: content/docs/practices/instrumentation.md [hunk 4]
```diff
@@ -206,11 +206,55 @@ Gauges can be set, go up, and go down. They are useful for snapshots of state,
...
+score](http://en.wikipedia.org/wiki/Apdex). Configure a bucket with the target
+request duration as upper bound and another bucket with 4 times the request
+duration as upper bound. Example: The target request duration is 250ms. The
+tolerable request duration is 1s. The request duration are collected with a
+histogram called `http_request_duration_seconds`. The following expression
+yields the Apdex score:
+
+```
+(
+  http_request_duration_seconds_bucket{le="0.25"} + http_request_duration_seconds_bucket{le="1"}
```
Comment (brian-brazil) [inline content/docs/practices/instrumentation.md]: It could be le="1.0" with some clients, not sure we should mention that.

We should be taking rates of these and aggregating up

File: content/docs/querying/functions.md [hunk 5]
```diff
@@ -28,6 +28,12 @@ the 1-element output vector from the input vector:
 This is useful for alerting on when no time series
 exist for a given metric name and label combination.
 
+## `bottomk()`
+
+`bottomk(k integer, v instant-vector` returns the `k` smallest elements of `v`
```
Comment (brian-brazil) [inline content/docs/querying/functions.md]: missing a )

File: content/docs/querying/functions.md [hunk 6]
```diff
@@ -80,6 +86,55 @@ and value across all series in the input vector.
 `floor(v instant-vector)` rounds the sample values of all elements in `v` down
 to the nearest integer.
 
+## `histogram_quantile()`
+
+`histogram_quantile(φ float, b instant-vector)` calculates the φ-quantile (0 ≤ φ
```
Comment (brian-brazil) [inline content/docs/querying/functions.md]: I think using phi might scare people off

File: content/docs/querying/functions.md [hunk 7]
```diff
@@ -80,6 +86,55 @@ and value across all series in the input vector.
...
+calculate the 90th percentile of request durations over the last 10m, use the
+following expression:
+
+```
+histogram_quantile(0.9, rate(http_request_duration_seconds_bucket[10m]))
+```
+
+The quantile is calculated for each label combination in
+`http_request_duration_seconds`. To aggregate, use the `sum()` aggregator
+outside of the `rate()` function. Since the `le` label is required by
```
Comment (brian-brazil) [inline content/docs/querying/functions.md]: s/outside/around/ - an example maybe? Probably belongs in the rules best practices.

File: content/docs/practices/instrumentation.md (hunk 1 above)
Comment (beorn7) [inline content/docs/practices/instrumentation.md]: In the next commit, I'll try to make the usual case clearer without denying the possibility of negative observations.

File: content/docs/practices/instrumentation.md (hunk 2 above)
Comment (beorn7) [inline content/docs/practices/instrumentation.md]: It's two orders of magnitude slower than rendering a time series straight... but the latter is very fast... I didn't want to make the table even more verbose so I didn't want to start a scientific discussion there about the exact costs...

File: content/docs/practices/instrumentation.md (hunk 3 above)
Comment (beorn7) [inline content/docs/practices/instrumentation.md]: I think we should not speculate on what the users want or need. Let's wait how the histograms work in practice, and with a bit of experience, we might make bolder statements here.

File: content/docs/practices/instrumentation.md (hunk 4 above)
Comment (beorn7) [inline content/docs/practices/instrumentation.md]: *gulp* (about the 1.0 vs. 1)

And fixed the rates.

File: content/docs/querying/functions.md (hunk 5 above)
Comment (beorn7) [inline content/docs/querying/functions.md]: Done.

File: content/docs/querying/functions.md (hunk 6 above)
Comment (beorn7) [inline content/docs/querying/functions.md]: It's the usual symbol for a quantile value. The problem is that usually, quantiles are named the "4-quantile" (for quartiles) and such. specifying the fraction, as we do, is not so common, but if it is done, it's oftes connected to φ. So I think it creates the least amount of confusion.

File: content/docs/querying/functions.md (hunk 7 above)
Comment (beorn7) [inline content/docs/querying/functions.md]: Done.

(I'll leave the best practice part to a different PR.)

File: content/docs/practices/instrumentation.md (hunk 2 above)
Comment (brian-brazil) [inline content/docs/practices/instrumentation.md]: The current wording makes it seem that calculating quantiles is very expensive, can we make it less so?

File: content/docs/practices/instrumentation.md (hunk 3 above)
Comment (brian-brazil) [inline content/docs/practices/instrumentation.md]: I'm going off my own experience working similar things in the past. You'd be surprised how much you can do with just a handful of buckets.

File: content/docs/querying/functions.md (hunk 6 above)
Comment (brian-brazil) [inline content/docs/querying/functions.md]: My concern is that most people are not good at math and are scared by anything that's more complex than addition and multiplication.

Using Greek letters is going to make many users discard this function (and possibly all of Prometheus) out of hand.

This documentation needs to not do that, we need to aim for the lowest-common denominator for at least the first paragraph.

Comment (juliusv) [inline content/docs/querying/functions.md]: I agree. Greek letters are "cool", but scary for the uninitiated, and I think very few Prometheus users will know the meaning of φ.

Comment (beorn7) [inline content/docs/querying/functions.md]: Happy to insert your suggestion here.

File: content/docs/practices/instrumentation.md (hunk 3 above)
Comment (beorn7) [inline content/docs/practices/instrumentation.md]: Sure. But I have to argue with 100+ Soundcloud engineers, some of them very opinionated. So I'll prefer not to open a discussion front here. Stating that you can't aggregate pre-calculated quantiles in a meaningful way is a much more convincing argument (because of... math!) then opening that can of worms in which situation you want which level of accuracy...

File: content/docs/practices/instrumentation.md (hunk 2 above)
Comment (beorn7) [inline content/docs/practices/instrumentation.md]: But it is pretty expensive. At least if you plot 2000 quantiles in one graph. You wouldn't believe what otherwise reasonable people do over here once they have the power of Prometheus in their mortal hands... :)

Comment (brian-brazil) [inline content/docs/practices/instrumentation.md]: Doing 2000 of anything is slow, that doesn't say much about histogram_quantile (I'd expect the timeseries lookups to dominate that particular query for example).
This language needs softening, or at least for this row to use the same language as the previous row (cheap vs expensive).

File: content/docs/practices/instrumentation.md [hunk 8]
```diff
@@ -206,11 +206,61 @@ Gauges can be set, go up, and go down. They are useful for snapshots of state,
...
+function](/docs/querying/functions/#histogram_quantile()).
+
+Both approaches have specific advantages and disadvantages:
+
+|   | Histogram | Summary
+|---|-----------|---------
+| Configuration | Need to configure buckets suitable for the expected range of observed values. | Need to configure φ-quantiles and sliding window, other φ-quantiles and sliding windows cannot be calculated later.
+| Client performance | Observations are very cheap as they only need to increment counters. | Observations are expensive due to the streaming quantile calculation.
+| Server performance | Calculating quantiles is expensive, consider [recording rules](/docs/querying/rules/#recording-rules) as a remedy. | Very low resource needs.
+| Number of time series | Low for Apdex score (see below), very high for accurate quantiles. Each bucket creates a time series. | Low, one time series per configured quantile.
```
Comment (brian-brazil) [inline content/docs/practices/instrumentation.md]: > very high for accurate quantiles

Citation needed.

"High if you have many buckets", accuracy is a separate question from precision.

File: content/docs/practices/instrumentation.md (hunk 3 above)
Comment (brian-brazil) [inline content/docs/practices/instrumentation.md]: I think we should generally aim our docs at new users, and encourage them down the right path where possible. 

What we want to avoid is users summarily choosing summaries as they think "perfect accuracy is critical", without understanding the broader implications or indeed that it isn't critical in most use cases (and if it is critical, you probably shouldn't be using Prometheus for it).

**Comment by beorn7** [comment]
I'll comment and rework the text ASAP. Just very busy with other stuff right now. Will get back to this tomorrow or so.

File: content/docs/concepts/metric_types.md [hunk 9]
```diff
@@ -49,26 +49,51 @@ See the client library usage documentation for gauges:
...
+  * the **count** of events that have been observed, exposed as `<basename>_count` (identical to `<basename>_bucket{le="+Inf"}` above)
+
+Use the [`histogram_quantile()`
+function](/docs/querying/functions/#histogram_quantile()) to calculate quantiles
+from histograms or even aggregations of histograms. A histogram is also suitable
+to calculate an [Apdex score](http://en.wikipedia.org/wiki/Apdex). See [summary
+vs. histogram](/docs/practices/instrumentation/#summary-vs.-histogram) for
+details of histogram usage and differences to [summaries](#summary).
+
+Client library usage documentation for summaries:
```
Comment (juliusv) [inline content/docs/concepts/metric_types.md]: s/for summaries/for histograms/

File: content/docs/concepts/metric_types.md [hunk 10]
```diff
@@ -49,26 +49,51 @@ See the client library usage documentation for gauges:
...
+
+   * [Go](http://godoc.org/github.com/prometheus/client_golang/prometheus#Histogram)
+   * [Java](https://github.com/prometheus/client_java/blob/master/simpleclient/src/main/java/io/prometheus/client/Histogram.java) (histograms are only supported by the simple client but not by the legacy client)
+   * [Python](https://github.com/prometheus/client_python#histogram)
+
+## Summary
+
+Similar to a _histogram_, a _summary_ samples observations (usually things like
+request durations and response sizes). While it also provides a total count of
+observation and a sum of all observed values, it calculates configurable
```
Comment (juliusv) [inline content/docs/concepts/metric_types.md]: s/observation/observations/

File: content/docs/practices/instrumentation.md (hunk 3 above)
Comment (juliusv) [inline content/docs/practices/instrumentation.md]: I agree with @brian-brazil here. Users will be looking for some kind of "guidance" to know what actually matters, and we might either scare them or send them on the wrong path otherwise.

File: content/docs/practices/instrumentation.md [hunk 11]
```diff
@@ -206,11 +206,61 @@ Gauges can be set, go up, and go down. They are useful for snapshots of state,
...
+probably collect request durations from every single instance in your fleet, and
+then you want to aggregate everything into an overall 95th percentile. You can
+only do that with histograms, but not with summaries. Aggregating the
+precomputed quantiles from a summary rarely makes sense.
+
+A histogram is suitable to calculate the [Apdex
+score](http://en.wikipedia.org/wiki/Apdex). Configure a bucket with the target
+request duration as upper bound and another bucket with 4 times the request
+duration as upper bound. Example: The target request duration is 250ms. The
+tolerable request duration is 1s. The request duration are collected with a
```
Comment (juliusv) [inline content/docs/practices/instrumentation.md]: s/duration/durations/

File: content/docs/querying/basics.md [hunk 12]
```diff
@@ -107,6 +107,30 @@ a `job` label set to `prometheus`:
...
 
+### Offset modifier
+
+The `offset` modifier allows changing the time offset for individual
+instant and range vectors in a query.
+
+For example, the following expression returns the value of `foo` 5
+minutes in the past relative to the current query evaluation time:
+
+    foo offset 5m
```
Comment (juliusv) [inline content/docs/querying/basics.md]: For the official docs, we might want to use a name other than `foo` that is more instantly recognizable as a metric name.

Like `api_http_requests_total`.

It would also be good to include a `{...}` label selector in at least one of these examples so that people see that that syntactical combination is possible: `api_http_requests_total{method="GET"} offset 5m`

Comment (juliusv) [inline content/docs/querying/basics.md]: (actually, s/api_http_requests_total/http_requests_total/, since that is the example metric used multiple times above.

File: content/docs/querying/functions.md (hunk 6 above)
Comment (juliusv) [inline content/docs/querying/functions.md]: Just `q` or `quantile`?

File: content/docs/querying/functions.md [hunk 13]
```diff
@@ -80,6 +86,55 @@ and value across all series in the input vector.
...
+histogram_quantile(0.9, sum(rate(http_request_duration_seconds_bucket[10m])) by (job, le))
+```
+
+To aggregate everything, specify only the `le` label:
+
+```
+histogram_quantile(0.9, sum(rate(http_request_duration_seconds_bucket[10m])) by (le))
+```
+
+The `histogram_quantile()` interpolates quantile values by assuming a linear
```
Comment (juliusv) [inline content/docs/querying/functions.md]: "The `histogram_quantile()` function"?

File: content/docs/practices/instrumentation.md (hunk 2 above)
Comment (beorn7) [inline content/docs/practices/instrumentation.md]: I'd expect that it will be dominated by quantile calculation. But that's not the point.

The point is that users will have done something heavy with summaries, and if they do the same with histograms, things will be slower. They need to be aware of that. The remedy is already in the table.

I have reworded the whole table anyway, though (wait for next commit). I aim to keep it factual but short (and neither soft nor harsh).

File: content/docs/practices/instrumentation.md (hunk 3 above)
Comment (beorn7) [inline content/docs/practices/instrumentation.md]: I would like to remind you that we already had bucketed histograms in a very early version of the client library. The whole shebang with precalculated quantiles stemmed from the bad experience with the bucketed approach. Histograms are hard to use correctly. We can't change that. By giving newbies some handwavy "guidance", we'll make things worse. Claiming that accuracy wrt buckets will not matter so much is like telling them, they could just do weighted average of their precalculated quantiles. The latter will also be "almost" correct in normal cases, and it will also give them an "indication of which direction it's moving". The big problem with all of that is that the "normal" cases are not the one where you have to rely on your monitoring. You need your monitoring when exploring exceptional cases, and that's exactly when the pointlessness of averaging quantiles strikes. Or, in terms of buckets, when all your relevant percentiles fall into the same bucket, and the distribution within the bucket is so uneven, that the linear interpolation we apply here is misleading at best.
Scientific ethics requires that we don't gloss over those relevant facts. (I feel a bit like Matt now... :)

In the next commit, you'll find a reworded version that goes into a bit more depth about those error cases.

File: content/docs/practices/instrumentation.md (hunk 4 above)
Comment (beorn7) [inline content/docs/practices/instrumentation.md]: I went down the loser's way and changed the values to avoid 1.0. We should keep this thing in mind and think about unified number formatting with client libraries.

File: content/docs/querying/functions.md (hunk 6 above)
Comment (beorn7) [inline content/docs/querying/functions.md]: As said, that will lure people down the more common way of a '4-quantile' (for 25th percentile). They will get weird results once they try to run `histogram_quantile(4, series)`.

Again, I don't think we can gain anything by pretending a hard thing is simple.

In the next commit, I add a section that explains φ-quantiles, hopefully in a way everybody understands it (or let's say only people will not understand it that should better not touch histograms at all).

File: content/docs/practices/instrumentation.md (hunk 8 above)
Comment (beorn7) [inline content/docs/practices/instrumentation.md]: This will be completely reworded in the next commit.

File: content/docs/concepts/metric_types.md (hunk 9 above)
Comment (beorn7) [inline content/docs/concepts/metric_types.md]: done.

File: content/docs/practices/instrumentation.md (hunk 11 above)
Comment (beorn7) [inline content/docs/practices/instrumentation.md]: Done.

File: content/docs/querying/functions.md (hunk 6 above)
Comment (juliusv) [inline content/docs/querying/functions.md]: Oooook.

File: content/docs/querying/basics.md (hunk 12 above)
Comment (beorn7) [inline content/docs/querying/basics.md]: Done.

File: content/docs/querying/functions.md (hunk 13 above)
Comment (beorn7) [inline content/docs/querying/functions.md]: Done.

**Comment by beorn7** [comment]
Still polishing.... commit might only happen tonight. Stay tuned. :)

**Comment by beorn7** [comment]
Sorry for the longer-then-expected delay... Here is the heavily reworked version.

Most importantly, I have created a completely new section for the delicate subject of histograms and
summaries and updated all the references.

File: content/docs/practices/histograms.md [hunk 14]
```diff
@@ -0,0 +1,220 @@
+---
+title: Histograms and summaries
+sort_rank: 4
+---
+
+# Histograms and summaries
+
+Histograms and summaries are more complex metric types. Not only
+creates a single histogram or summary a multitude of time series, it
```
Comment (brian-brazil) [inline content/docs/practices/histograms.md]: Grammar is off here

File: content/docs/practices/histograms.md [hunk 15]
```diff
@@ -0,0 +1,220 @@
...
+
+Histograms and summaries both sample observations, typically request
+durations or response sizes. They track the number of observations
+*and* the sum of the observed values, allowing you to calculate the
+*average* of the observed values. Note that the number of observations
+(showing up in Prometheus as a time series with a `_count` suffix) is
+inherently a counter (as described above, it only goes up). The sum of
+observations (showing up as a time series with a `_sum` suffix)
+behaves like a counter, too, as long as all observations are
+positive. Obviously, request durations or response sizes are always
```
Comment (brian-brazil) [inline content/docs/practices/histograms.md]: s/positive/not negative/

File: content/docs/practices/histograms.md [hunk 16]
```diff
@@ -0,0 +1,220 @@
...
+to observe negative values (e.g. temperatures in centigrade). In that
+case, the sum of observations can go down, so you cannot apply
+`rate()` to it anymore.
+
+To calculate the average request duration during the last 5 minutes
+from a histogram or summary called `http_request_duration_second`, use
+the following expression:
+
+    rate(http_request_duration_seconds_sum[5m])
+      /
```
Comment (brian-brazil) [inline content/docs/practices/histograms.md]: This formatting will be hard to read when nested, use:
  a
/
  b

File: content/docs/practices/histograms.md [hunk 17]
```diff
@@ -0,0 +1,220 @@
...
+    sum(rate(http_request_duration_seconds_count[5m])) by (job)
+
+
+You can calculate the well-known [Apdex
+score](http://en.wikipedia.org/wiki/Apdex) in a similar way. Configure
+a bucket with the target request duration as upper bound and another
+bucket with the tolerated request duration (usually 4 times the target
+request duration) as upper bound. Example: The target request duration
+is 300ms. The tolerable request duration is 1.2s. The following
+expression yields the Apdex score over the last 5 minutes:
```
Comment (brian-brazil) [inline content/docs/practices/histograms.md]: The above example is across all instances, this one is per instance. We should probably have both across all.

File: content/docs/practices/histograms.md [hunk 18]
```diff
@@ -0,0 +1,220 @@
...
+will fall into the bucket labeled `{le="0.3"}`, i.e. the bucket from
+200ms to 300ms. The histogram implementation guarantees that the true
+95th percentile is somewhere between 100ms and 200ms. To return a
+single value (rather than an interval), it applies linear
+interpolation, which yields 295ms in this case. The calculated
+quantile gives you the impression that you are close to breaking the
+SLA, but in reality, the 95th percentile is a tiny bit above 220ms,
+a quite comfortable distance to your SLA.
+
+Next step in our *Gedenkenexperiment*: A change in backend routing
```
Comment (brian-brazil) [inline content/docs/practices/histograms.md]: You seem to have accidentally slipped into German here :)

File: content/docs/practices/histograms.md [hunk 19]
```diff
@@ -0,0 +1,220 @@
...
+dimension of the observed value (via choosing the appropriate bucket
+layout). With a broad distribution, small changes in φ result in
+large deviations in the observed value. With a sharp distribution, a
+small interval of observed values covers a large interval of φ.
+
+Two rules of thumb:
+
+  1. If you need to aggregate, choose histograms.
+
+  2. Otherwise, choose a histogram if you need accuracy in the
```
Comment (brian-brazil) [inline content/docs/practices/histograms.md]: This is a bit wordy for a rule of thumb.

Maybe something along the lines of:
choose a histogram if you have an idea of the range/distribution of values that will be seen and have a good idea of the buckets to use. Choose a summary if you need an accurate quartile, no matter what the range/distribution of the values is.

File: content/docs/querying/functions.md [hunk 20]
```diff
@@ -80,6 +86,54 @@ and value across all series in the input vector.
 `floor(v instant-vector)` rounds the sample values of all elements in `v` down
 to the nearest integer.
 
+## `histogram_quantile()`
+
+`histogram_quantile(φ float, b instant-vector)` calculates the
```
Comment (brian-brazil) [inline content/docs/querying/functions.md]: Can we drop the greek letter from here please? 'q' or 'quantile' should do.

File: content/docs/practices/histograms.md [hunk 21]
```diff
@@ -0,0 +1,220 @@
...
+
+## Library support
+
+First of all, check the library support for
+[histograms](/docs/concepts/metric_types/#histogram) and
+[summaries](/docs/concepts/metric_types/#summary). Full support for
+both currently only exists in the Go client library. Many libraries
+support only one of the two types, or they support summaries only in a
+limited fashion (lacking [quantile
+calculation](#quantiles)). [Contributions are welcome](/community/),
```
Comment (brian-brazil) [inline content/docs/practices/histograms.md]: The historical note and call for contributions feels off at the start of the doc, move to the end maybe?

File: content/docs/practices/histograms.md (hunk 14 above)
Comment (beorn7) [inline content/docs/practices/histograms.md]: In which way? Need language assistance...

File: content/docs/practices/histograms.md (hunk 15 above)
Comment (beorn7) [inline content/docs/practices/histograms.md]: Done.

File: content/docs/practices/histograms.md (hunk 16 above)
Comment (beorn7) [inline content/docs/practices/histograms.md]: Thanks. That's exactly the advice I was fishing for.

File: content/docs/practices/histograms.md (hunk 17 above)
Comment (beorn7) [inline content/docs/practices/histograms.md]: Done.

File: content/docs/practices/histograms.md (hunk 18 above)
Comment (beorn7) [inline content/docs/practices/histograms.md]: Not accidentally.

http://www.merriam-webster.com/dictionary/gedankenexperiment

It's very physics... if that feels off too much, I'm happy tho change it.

File: content/docs/practices/histograms.md (hunk 19 above)
Comment (beorn7) [inline content/docs/practices/histograms.md]: Gave it another try.

File: content/docs/querying/functions.md (hunk 20 above)
Comment (beorn7) [inline content/docs/querying/functions.md]: Surely we can do that, but then we have to write a line further down that `d` stands for the φ in "φ-quantile". It's an additional layer of indirection. I can't see how that will help to reduce confusion.

File: content/docs/practices/histograms.md (hunk 21 above)
Comment (beorn7) [inline content/docs/practices/histograms.md]: Done.

File: content/docs/practices/histograms.md (hunk 18 above)
Comment (juliusv) [inline content/docs/practices/histograms.md]: It's "Gedankenexperiment" though, not "Gedenkenexperiment" :)

But yeah - please strive for maximum clarity in docs, not maximum witfulness. "Thought experiment" or something like that will be easier to understand for most people.

File: content/docs/practices/histograms.md (hunk 14 above)
Comment (juliusv) [inline content/docs/practices/histograms.md]: "Not only creates a single histogram" is missing a subject, or at least it seems that way, because I think you mean "histogram or summary" to be the subject, but the location they appear in makes them seem like the object.

Probably you want to say "Not only do they create a multitude of time series, it is also ..."

File: content/docs/practices/histograms.md [hunk 22]
```diff
@@ -0,0 +1,224 @@
...
+observations (showing up as a time series with a `_sum` suffix)
+behaves like a counter, too, as long as there are no negative
+observations. Obviously, request durations or response sizes are
+never negative. In principle, however, you can use summaries and
+histograms to observe negative values (e.g. temperatures in
+centigrade). In that case, the sum of observations can go down, so you
+cannot apply `rate()` to it anymore.
+
+To calculate the average request duration during the last 5 minutes
+from a histogram or summary called `http_request_duration_second`, use
```
Comment (juliusv) [inline content/docs/practices/histograms.md]: s/http_request_duration_second/http_request_duration_seconds/

File: content/docs/practices/histograms.md [hunk 23]
```diff
@@ -0,0 +1,224 @@
...
+a histogram called `http_request_duration_seconds`.
+
+      sum(rate(http_request_duration_seconds_bucket{le="0.3"}[5m])) by (job)
+    /
+      sum(rate(http_request_duration_seconds_count[5m])) by (job)
+
+
+You can calculate the well-known [Apdex
+score](http://en.wikipedia.org/wiki/Apdex) in a similar way. Configure
+a bucket with the target request duration as upper bound and another
```
Comment (juliusv) [inline content/docs/practices/histograms.md]: "as upper bound" -> "as the upper bound"

File: content/docs/practices/histograms.md [hunk 24]
```diff
@@ -0,0 +1,224 @@
...
+bucket with the tolerated request duration (usually 4 times the target
+request duration) as upper bound. Example: The target request duration
```
Comment (juliusv) [inline content/docs/practices/histograms.md]: Dito.

File: content/docs/practices/histograms.md [hunk 25]
```diff
@@ -0,0 +1,224 @@
...
+## Quantiles
+
+You can use both summaries and histograms to calculate so-called φ-quantiles,
+where 0 ≤ φ ≤ 1. The φ-quantile is the observation value that ranks at number
+φ*N among the N observations. Examples for φ-quantiles: The 0.5-quantile is
+known as the median. The 0.95-quantile is the 95th percentile.
+
+The essential difference between summaries and histograms is that summaries
+calculate streaming φ-quantiles on the client side and expose them directly,
+while histograms expose bucketed observations counts and the calculation of
```
Comment (juliusv) [inline content/docs/practices/histograms.md]: "observations counts" -> "observation counts"

File: content/docs/practices/histograms.md [hunk 26]
```diff
@@ -0,0 +1,224 @@
...
+| Specification of φ-quantile and sliding time-window | Ad-hoc with [Prometheus expressions](/docs/querying/functions/#histogram_quantile()). | Preconfigured by the client.
+| Aggregation | Ad-hoc with [Prometheus expressions](/docs/querying/functions/#histogram_quantile()). | In general [not aggregatable](http://latencytipoftheday.blogspot.de/2014/06/latencytipoftheday-you-cant-average.html).
+
+Note the importance of the last item in the table. Let us return to
+the SLA of serving 95% of requests within 300ms. This time, you do not
+want to display the percentage of requests served within 300ms, but
+instead the 95th percentile, i.e. the request duration within which
+you have served 95% of requests. To do that, you can either configure
+a summary with a 0.95-quantile and (for example) a 5-minute decay
+time-window, or you configure a histogram with a few buckets around
```
Comment (juliusv) [inline content/docs/practices/histograms.md]: "time-window" -> "time window"

File: content/docs/practices/histograms.md (hunk 14 above)
Comment (beorn7) [inline content/docs/practices/histograms.md]: I see, the "do" is missing. http://dictionary.cambridge.org/grammar/british-grammar/not-only-but-also

File: content/docs/practices/histograms.md [hunk 27]
```diff
@@ -0,0 +1,224 @@
...
+200ms to 300ms. The histogram implementation guarantees that the true
+95th percentile is somewhere between 100ms and 200ms. To return a
+single value (rather than an interval), it applies linear
+interpolation, which yields 295ms in this case. The calculated
+quantile gives you the impression that you are close to breaking the
+SLA, but in reality, the 95th percentile is a tiny bit above 220ms,
+a quite comfortable distance to your SLA.
+
+Next step in our *Gedenkenexperiment*: A change in backend routing
+adds a fixed amount of 100ms to all requent durations. Now the request
```
Comment (juliusv) [inline content/docs/practices/histograms.md]: "requent" -> "request"

File: content/docs/practices/histograms.md [hunk 28]
```diff
@@ -0,0 +1,224 @@
...
+duration has its sharp spike at 320ms and almost all observations will
+fall into the bucket from 300ms to 450ms. The 95th percentile is
+calculated to be 442.5ms, although the correct values is close to
```
Comment (juliusv) [inline content/docs/practices/histograms.md]: "values" -> "value"

File: content/docs/practices/histograms.md [hunk 29]
```diff
@@ -0,0 +1,224 @@
...
+
+
+## What can I do if my client library does not support the metric type I need?
+
+Implement it! [Code contributions are welcome](/community/). In
+general, we expect histograms to be more urgently needed than
+summaries. Histograms are also easier to implement in a client
+library, so we recommend to implement histograms first, if in
+doubt. The reason why some libraries offer summaries but not
+histograms (Ruby, the legacy Java client) is that histograms are a
```
Comment (juliusv) [inline content/docs/practices/histograms.md]: "," -> " and"

File: content/docs/querying/functions.md (hunk 20 above)
Comment (juliusv) [inline content/docs/querying/functions.md]: Given that we had this discussion last week (and no clear outcome), I think it's ok to keep it as it is for now.

File: content/docs/practices/histograms.md (hunk 14 above)
Comment (juliusv) [inline content/docs/practices/histograms.md]: Well, also the subject is in the wrong position, and the verb is not conjugated correctly? I would just copy my suggestion :)

Comment (beorn7) [inline content/docs/practices/histograms.md]: "a single histogram or summary" is clearly singular. A "does" was missing, and then "creates" becomes infinite and goes to a different place. The subject was correctly placed from the beginning.

Should be good now.

File: content/docs/practices/histograms.md (hunk 18 above)
Comment (beorn7) [inline content/docs/practices/histograms.md]: Removed the nerdiness.

File: content/docs/practices/histograms.md (hunk 22 above)
Comment (beorn7) [inline content/docs/practices/histograms.md]: Done.

File: content/docs/practices/histograms.md (hunk 23 above)
Comment (beorn7) [inline content/docs/practices/histograms.md]: Done.

File: content/docs/practices/histograms.md (hunk 24 above)
Comment (beorn7) [inline content/docs/practices/histograms.md]: Done.

File: content/docs/practices/histograms.md (hunk 25 above)
Comment (beorn7) [inline content/docs/practices/histograms.md]: DOne.

File: content/docs/practices/histograms.md (hunk 26 above)
Comment (beorn7) [inline content/docs/practices/histograms.md]: I thought it's a time window for the decay rather than a window of decay time.

I guess it's clearer anyway to call it a '5-minute decay time' without the window.

File: content/docs/practices/histograms.md (hunk 27 above)
Comment (beorn7) [inline content/docs/practices/histograms.md]: Done.

File: content/docs/practices/histograms.md (hunk 28 above)
Comment (beorn7) [inline content/docs/practices/histograms.md]: Doen.

File: content/docs/practices/histograms.md (hunk 29 above)
Comment (beorn7) [inline content/docs/practices/histograms.md]: DOne

**Comment by beorn7** [comment]
PHAL

**Comment by juliusv** [comment]
:+1:
